pkg/kubelet/cm/memorymanager/memory_manager.go

```diff
@@ -147,5 +249,83 @@ func (m *manager) State() state.Reader {
+}
+
+// TODO: consider to move this method to manager interface, the only difference between CPU manager is assignments, we can send it to the method
+func (m *manager) removeStaleState() {
+	// Only once all sources are ready do we attempt to remove any stale state.
+	// This ensures that the call to `m.activePods()` below will succeed with
+	// the actual active pods list.
+	if !m.sourcesReady.AllReady() {
+		return
+	}
+
+	// We grab the lock to ensure that no new containers will grab memory block while
+	// executing the code below. Without this lock, its possible that we end up
+	// removing state that is newly added by an asynchronous call to
+	// AddContainer() during the execution of this code.
+	m.Lock()
+	defer m.Unlock()
+
+	// Get the list of active pods.
+	activePods := m.activePods()
+
+	// Build a list of (podUID, containerName) pairs for all containers in all active Pods.
+	activeContainers := make(map[string]map[string]struct{})
+	for _, pod := range activePods {
+		activeContainers[string(pod.UID)] = make(map[string]struct{})
+		for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+			activeContainers[string(pod.UID)][container.Name] = struct{}{}
+		}
+	}
+
+	// Loop through the MemoryManager state. Remove any state for containers not
+	// in the `activeContainers` list built above.
+	assignments := m.state.GetMemoryAssignments()
+	for podUID := range assignments {
+		for containerName := range assignments[podUID] {
+			if _, ok := activeContainers[podUID][containerName]; !ok {
+				klog.Infof("[memorymanager] removeStaleState: removing (pod %s, container: %s)", podUID, containerName)
+				err := m.policyRemoveContainerByRef(podUID, containerName)
+				if err != nil {
+					klog.Errorf("[memorymanager] removeStaleState: failed to remove (pod %s, container %s), error: %v)", podUID, containerName, err)
+				}
+			}
+		}
+	}
+}
+
+func (m *manager) policyRemoveContainerByRef(podUID string, containerName string) error {
+	err := m.policy.RemoveContainer(m.state, podUID, containerName)
+	if err == nil {
+		m.containerMap.RemoveByContainerRef(podUID, containerName)
+	}
+
+	return err
+}
 
+func getReservedMemory(machineInfo *cadvisorapi.MachineInfo, nodeAllocatableReservation v1.ResourceList) (systemReservedMemory, error) {
+	// TODO: we should add new kubelet parameter, and to get reserved memory per NUMA node from it
+	// currently we use kube-reserved + system-reserved + eviction reserve for each NUMA node, that creates memory over-consumption
+	// and no reservation for huge pages
+	reserved := systemReservedMemory{}
+	for _, node := range machineInfo.Topology {
+		memory := nodeAllocatableReservation[v1.ResourceMemory]
+		if memory.IsZero() {
+			break
+		}
+		value, succeeded := memory.AsInt64()
+		if !succeeded {
+			return nil, fmt.Errorf("failed to represent reserved memory as int64")
+		}
+
+		reserved[node.Id] = map[v1.ResourceName]uint64{
+			v1.ResourceMemory: uint64(value),
+		}
+	}
```

Comment:
Is this actually correct? Every NUMA node has `nodeAllocatableReservation[v1.ResourceMemory]` associated with it?

I would have thought that `nodeAllocatableReservation[v1.ResourceMemory]` represents the total memory on the machine, not the total memory per NUMA node.

Reply:
You right, you will see the change after the commit that will introduce the memory manager `--reserved-memory` flag.